quickstarts/microsoft.keyvault/key-vault-managed-identity-role-assignment/main.bicep

```diff
@@ -0,0 +1,40 @@
+@description('Specifies the name of the key vault.')
+param keyVaultName string
+
+@description('Specifies the SKU to use for the key vault.')
+param keyVaultSku object
+
+@description('Specifies the Azure location where the resources should be created.')
+param location string = resourceGroup().location
+
+var managedIdentityName = 'my-managed-identity'
+
+resource keyVault 'Microsoft.KeyVault/vaults@2021-10-01' = {
+  name: keyVaultName
+  location: location
+  properties: {
+    enableRbacAuthorization: true
+    tenantId: tenant().tenantId
+    sku: keyVaultSku
+  }
+}
+
+resource managedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' = {
+  name: managedIdentityName
+  location: location
+}
+
+@description('This is the built-in Key Vault Administrator role. See https://docs.microsoft.com/azure/role-based-access-control/built-in-roles#key-vault-administrator')
+resource keyVaultAdministratorRoleDefinition 'Microsoft.Authorization/roleDefinitions@2018-01-01-preview' existing = {
+  scope: subscription()
+  name: '00482a5a-887f-4fb3-b363-3b7fe8e74483'
+}
+
+module roleAssignment 'modules/role-assignment.bicep' = {
```

Comment:
For this specific example, I'd say yes. I've added an explanation to the readme, but essentially this example is in response to some customer feedback about how difficult it is to deploy role assignments and specifically how they behave when the principal (e.g. a managed identity) is deleted and then you try to redeploy the resources. Because the new managed identity's principal ID is different, the redeployment of the role assignment fails. To work around this, we can use the principal ID as one of the seeds to the `guid()` function for the role assignment ID, but when you do this you must use a module.

Reply:
Ah, got it... so we should be able to use:

```
guid(resourceGroup().id, managedIdentity.id, keyVaultAdministratorRoleDefinition.id)
```

and use a single file...

Also, it seems like the scope should be the vault and not the RG?

Reply:
I've updated the scope.

Re the name - for this specific example, we want to use the `managedIdentity.properties.principalId` rather than `managedIdentity.id` because the resource ID will be the same on a redeployment (after the MI is deleted) but the principal ID is different, which is where this problem comes about. So I think we do need to leave the name as-is and use a module.

Reply:
if managedIdentity is not idempotent?